### ISSUE TITLE
## type

### ISSUE THREAD

**Issue by Nigel-Galloway** [pascalabcnet/pascalabcnet#3209]
While attempting to translate my F# entry to PascalABC.NET at https://rosettacode.org/wiki/I_before_E_except_after_C#PascalABC.NET I wished to write

'##
type plausibility=(plausible,implausible);'

the compiler responded "Found 'type' but expected operator". If I can not use type when using ## it severely restricts the vale of ##.

**Comment by SunSerega** [comment]
> '##
> type plausibility=(plausible,implausible);'

This is still not the [Markdown syntax](https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).

When pasted properly, it would look like this:
```
##
type plausibility=(plausible,implausible);
```

---

The common explanation by the devs is that `##` is for short programs, and programs where you need custom types are too complex and would become a mess, if type declaration is allowed.
Personally, I think this is too arbitrary of a bound, and the user should decide. But your example is actually showing that I'm wrong...
```
// I before E except after C. Nigel Galloway: September 30th., 2024
type plausibility=(plausible,implausible);
var fN:string->string->integer:=n->g->System.Text.RegularExpressions.Regex.Matches(n,g).count;
var fG:Func<string,integer>->(array of integer)->array of integer:=n->g->g.Zip(|n('ie'),n('ei'),n('cie'),n('cei')|,(n,g)->n+g).ToArray;
begin
  var n:=System.IO.File.ReadLines('unixdict.txt').Aggregate(|0,0,0,0|,(n,g)->fG(fN(g))(n));
  println($'I before E except after C is {if n[0]-n[2]>2*n[2] then plausible else implausible}');
  println($'E before I except after C is {if n[1]-n[3]>2*n[3] then plausible else implausible}');
  println($'I before E   when after C is {if n[2]>2*n[3] then plausible else implausible}');
  println($'E before I   when after C is {if n[3]>2*n[2] then plausible else implausible}');
end.
```
Here, you don't actually use `plausibility`: the type; you only use 2 constants you declared together with it.

Instead, declaring:
```
var (plausible, implausible) := ('plausible', 'implausible');
```
Would achieve the same effect. It would also be more direct and not an abuse of an unrelated language feature.

Or, if your goal is to write a shorter code, no matter how ugly and silly:
```
var p: (plausible, implausible);
```
This is an anonymous type declaration embedded in a variable declaration. The variable is not used, the same way your `plausibility` type isn't.

But I think the best option would be:
```
##

var plausibility := Dict(true to 'plausible', false to 'implausible');

$'1. {plausibility[1>2]}'.Println;
$'2. {plausibility[2>1]}'.Println;
```

---

TL;DR: If you want to also push against this random restriction, you need a use case that is at least less random than the restriction...

**Comment by Nigel-Galloway** [comment]
I am happy using begin and end. rather than ##.  Miks1965 added an alternative version to one of my entries using ## claiming it is more modern, I wouldn't want to be anything but modern! I wanted to use ## here because I noticed that issue 3204 goes away then. I wanted to use type because the underlying logic here is:
```
myFun:Func<integer,integer,boolean>->plausibility
```
The options you suggest are more or less complicated ways of writing:
```
myFun:Func<integer,integer,boolean>->string
```
Doing everything with strings is bad practice, the type is more modern I think.

The code as is will not compile with PascalABC.NET not greater than 3.10, I am unconvinced that ## as is is useful if I can not just change between the two.

**Comment by SunSerega** [comment]
Replacing:
```
##
function check1(i: integer) := i.IsOdd;

ReadArrInteger(10).Where(check1).Println;
```
... with:
```
##
ReadArrInteger(10).Where(i->i.IsOdd).Println;
```
... is more modern because here, we need a single-use sub-program, and we don't need it to have a particular name.
But it doesn't mean replacing all sub-programs with lambdas will make them more modern.

In the same way, `##` is not meant to be a new way to write all programs - it's a short program header.

**Comment by miks1965** [comment]
Unfortunately, not all constructs can be used in programs with ##. This is due to specific grammar features. type is one of such constructs, unfortunately.
